SmartDeviceLink/SDLChoiceSetManager.m

```diff
@@ -0,0 +1,402 @@
+        if (handler != nil) {
+            handler(nil);
+        }
+
+        return;
+    }
+
+    [self sdl_updateIdsOnChoices:choicesToUpload];
+
+    // Add the preload cells to the pending preloads
+    [self.pendingMutablePreloadChoices unionSet:choicesToUpload];
+
+    // Upload pending preloads
+    SDLPreloadChoicesOperation *preloadOp = [[SDLPreloadChoicesOperation alloc] initWithConnectionManager:self.connectionManager fileManager:self.fileManager displayCapabilities:self.displayCapabilities isVROptional:self.isVROptional cellsToPreload:choicesToUpload];
+
+    __weak typeof(self) weakSelf = self;
+    __weak typeof(preloadOp) weakPreloadOp = preloadOp;
+    preloadOp.completionBlock = ^{
+        [weakSelf.preloadedMutableChoices unionSet:choicesToUpload];
+        [weakSelf.pendingMutablePreloadChoices minusSet:choicesToUpload];
+
+        if (handler != nil) {
+            handler(weakPreloadOp.error);
+        }
+    };
+    [self.transactionQueue addOperation:preloadOp];
+}
+
+- (void)deleteChoices:(NSArray<SDLChoiceCell *> *)choices {
+    if (![self.currentState isEqualToString:SDLChoiceManagerStateReady]) { return; }
+
+    // Find cells to be deleted that are already uploaded or are pending upload
+    NSSet<SDLChoiceCell *> *cellsToBeDeleted = [self sdl_choicesToBeDeletedWithArray:choices];
+    NSSet<SDLChoiceCell *> *cellsToBeRemovedFromPending = [self sdl_choicesToBeRemovedFromPendingWithArray:choices];
+
+    // If choices are deleted that are already uploaded or pending and are used by a pending presentation, cancel it and send an error
+    NSSet<SDLChoiceCell *> *pendingPresentationSet = [NSSet setWithArray:self.pendingPresentationSet.choices];
+    if ([cellsToBeDeleted intersectsSet:pendingPresentationSet] || [cellsToBeRemovedFromPending intersectsSet:pendingPresentationSet]) {
+        if (self.pendingPresentationSet.delegate != nil) {
+            [self.pendingPresentationSet.delegate choiceSet:self.pendingPresentationSet didReceiveError:[NSError sdl_choiceSetManager_choicesDeletedBeforePresentation:@{@"deletedChoices": choices}]];
+        }
+
+        self.pendingPresentationSet = nil;
+    }
+
+    // Remove the cells from pending and delete choices
+    [self.pendingMutablePreloadChoices minusSet:cellsToBeRemovedFromPending];
+    SDLDeleteChoicesOperation *deleteOp = [[SDLDeleteChoicesOperation alloc] initWithConnectionManager:self.connectionManager cellsToDelete:cellsToBeDeleted];
+
+    __weak typeof(self) weakself = self;
+    __weak typeof(deleteOp) weakOp = deleteOp;
+    deleteOp.completionBlock = ^{
+        if (weakOp.error != nil) {
+            SDLLogE(@"Failed to delete choices: %@", weakOp.error);
+            return;
+        }
+
+        [weakself.preloadedMutableChoices minusSet:cellsToBeDeleted];
+    };
+    [self.transactionQueue addOperation:deleteOp];
+}
+
+#pragma mark Present
+
+- (void)presentChoiceSet:(SDLChoiceSet *)choiceSet mode:(SDLInteractionMode)mode {
```

Comment:
Should do some data validation here. 
* There can't be more than 100 items in a choice set.
* Initial text can't be more than 500 characters.
* The prompts can't be empty arrays
* The timeout can't be greater than 100,000 milliseconds.
Sending any of the above will result in an `INVALID_DATA` response. Instead of sending a RPC we know will fail, we should just return an error with a detailed description of what is wrong.

Reply:
All done by returning nil from the `SDLChoiceSet`  initializer